src/ca.c

```diff
@@ -2736,625 +2419,770 @@ static int ciplus_app_upgr_message(void *arg, uint8_t slot_id,
+    DEFAPP(TS103205_APP_OPRF_TWO_RESOURCEID, CIPLUS_APP_OPRF_handler, NULL),
+    DEFAPP(TS103205_APP_OPRF_THREE_RESOURCEID, CIPLUS_APP_OPRF_handler, NULL),
+    // SAS
+    DEFAPP(CIPLUS_APP_SAS_RESOURCEID, CIPLUS_APP_SAS_handler, APP_SAS_create),
+    // Application MMI
+    DEFAPP(TS101699_APP_AMMI_RESOURCEID, APP_empty, NULL),
+    DEFAPP(CIPLUS_APP_AMMI_RESOURCEID, APP_empty, NULL),
+    // TS103205_APP_AMMI_RESOURCEID, // Multi-stream
+    // Multi-stream capability,
+    // TS103205_APP_MULTISTREAM_RESOURCEID
+};
 
-    return 0;
+int find_session_for_resource(int resource) {
+    int i;
+    for (i = 1;
+         i < sizeof(application_handler) / sizeof(application_handler[0]); i++)
+        if (resource == application_handler[i].resource)
+            return i;
+    return -1;
 }
 
-static int ca_mmi_display_ctl_callback(void *arg, uint8_t slot_id,
-                                       uint16_t session_number, uint8_t cmd_id,
-                                       uint8_t mmi_mode) {
-    ca_device_t *d = arg;
+int populate_resources(ca_device_t *d, int *resource_ids) {
+    int i;
+    for (i = 1;
+         i < sizeof(application_handler) / sizeof(application_handler[0]);
+         i++) {
+        int resource = application_handler[i].resource;
+        if (d->force_ci && resource == CIPLUS_APP_CC_RESOURCEID) {
+            break;
+        }
+        resource_ids[i - 1] = htonl(resource);
+    }
+    return i - 1;
+}
 
-    LOG("mmi display ctl cb received for slot %u session_num %u "
-        "cmd_id 0x%02x mmi_mode %u",
-        slot_id, session_number, cmd_id, mmi_mode);
+/* from dvb-apps */
+int asn_1_decode(int *length, unsigned char *asn_1_array) {
+    uint8_t length_field;
 
-    if (cmd_id == MMI_DISPLAY_CONTROL_CMD_ID_SET_MMI_MODE) {
-        struct en50221_app_mmi_display_reply_details det;
+    length_field = asn_1_array[0];
 
-        det.u.mode_ack.mmi_mode = mmi_mode;
-        if (en50221_app_mmi_display_reply(d->mmi_resource, session_number,
-                                          MMI_DISPLAY_REPLY_ID_MMI_MODE_ACK,
-                                          &det)) {
-            LOG("Slot %u: Failed to send MMI mode ack reply", slot_id);
-        }
+    if (length_field < 0x80) {
+        // there is only one word
+        *length = length_field & 0x7f;
+        return 1;
+    } else if (length_field == 0x81) {
+        *length = asn_1_array[1];
+        return 2;
+    } else if (length_field == 0x82) {
+        *length = (asn_1_array[1] << 8) | asn_1_array[2];
+        return 3;
     }
 
-    return 0;
+    return -1;
 }
 
-static int ca_mmi_enq_callback(void *arg, uint8_t slot_id,
-                               uint16_t session_number, uint8_t blind_answ,
-                               uint8_t exp_answ_len, uint8_t *text,
-                               uint32_t text_size) {
-    ca_device_t *d = arg;
-    char buffer[256];
+int asn_1_encode(int length, uint8_t *asn_1_array) {
+    if (length < 0x80) {
```

Comment:
Incidentally this fixes a bug in the previous implementation (was `< 127` when it should have been `< 128`). I had a fix for this coming up at some point.

Reply:
yeah I saw multiple implementation each with different logic... so I took the one from dvb-apps.